### PR TITLE
Allow Anonymous Access to Apps

### PR DESCRIPTION
**Pull request by lahiru823**
fix to redmine issue:
https://redmine.wso2.com/issues/3539

### CODE REVIEW

File: modules/apps/publisher/modules/ext/scripts/webapp.action.save.js [hunk 1]
```diff
@@ -47,7 +47,7 @@ var module = function () {
 	}
 
     //TODO: Change this method to take WebAppObj as argument instead of passing properties separately.
-    function addToWebApp(uuid,webappProvider, webappName, webappVersion, webappContext, webappTrackingCode,asset, ssoEnabled, idpProviderUrl, saml2SsoIssuer, logoutURL) {
+    function addToWebApp(uuid,webappProvider, webappName, webappVersion, webappContext, webappTrackingCode,asset, ssoEnabled, idpProviderUrl, saml2SsoIssuer, logoutURL,allowAnonymous) {
```
Comment (rushmin) [inline modules/apps/publisher/modules/ext/scripts/webapp.action.save.js]: The method accepts too many parameters. Consider using a wrapper object.

File: modules/apps/publisher/modules/ext/scripts/webapp.action.save.js [hunk 2]
```diff
@@ -157,6 +168,7 @@ var module = function () {
             var name = model.getField('overview.name').value;
             var version = model.getField('overview.version').value;
             var contextname = model.getField('overview.context').value;
+            var allowAnonymous=model.getField('overview.allowAnonymous').value;
```
Comment (rushmin) [inline modules/apps/publisher/modules/ext/scripts/webapp.action.save.js]: Need spaces !

File: modules/apps/publisher/themes/appm/js/copy-app.js [hunk 3]
```diff
@@ -332,6 +332,7 @@ $(function() {
         var app_transport = $('#overview_transports').val();
         var app_context = $('#overview_context').val();
         var app_provider = $('#overview_provider').val();
+        var app_allowAnonymous=$('#overview_allowAnonymous').val();
```
Comment (rushmin) [inline modules/apps/publisher/themes/appm/js/copy-app.js]: Need spaces !

File: modules/apps/publisher/themes/appm/js/copy-app.js [hunk 4]
```diff
@@ -353,6 +354,7 @@ $(function() {
         sso_config.app_transport = app_transport;
         sso_config.app_context = app_context;
         sso_config.app_provider = app_provider;
+        sso_config.app_allowAnonymous=app_allowAnonymous;
```
Comment (rushmin) [inline modules/apps/publisher/themes/appm/js/copy-app.js]: Need spaces !

File: modules/apps/publisher/themes/appm/js/edit.asset.js [hunk 5]
```diff
@@ -460,6 +460,7 @@ $(function() {
         	var app_version = $('#overview_version').val();
         	var app_transport = $('#overview_transports').val();
         	var app_context = $('#overview_context').val();
+			var app_allowAnonymous=$('#overview_allowAnonymous').val();
```
Comment (rushmin) [inline modules/apps/publisher/themes/appm/js/edit.asset.js]: Formatting issue.

File: modules/apps/publisher/themes/appm/js/edit.asset.js [hunk 6]
```diff
@@ -481,6 +482,7 @@ $(function() {
 		sso_config.app_verison = app_version;
 		sso_config.app_transport = app_transport;
 		sso_config.app_context = app_context;
+		sso_config.app_allowAnonymous=app_allowAnonymous;
```
Comment (rushmin) [inline modules/apps/publisher/themes/appm/js/edit.asset.js]: Need spaces !

File: modules/apps/publisher/themes/appm/js/logic/create.asset.js [hunk 7]
```diff
@@ -368,6 +368,7 @@ $(function() {
 		var app_transport = $('#overview_transports').val();
 		var app_context = $('#overview_context').val();
 		var app_provider = $('#overview_provider').val();
+		var app_allowAnonymous=$('#overview_allowAnonymous').val();
```
Comment (rushmin) [inline modules/apps/publisher/themes/appm/js/logic/create.asset.js]: Need spaces !

File: modules/apps/publisher/themes/appm/js/logic/create.asset.js [hunk 8]
```diff
@@ -389,6 +390,7 @@ $(function() {
 		sso_config.app_transport = app_transport;
 		sso_config.app_context = app_context;
 		sso_config.app_provider = app_provider;
+		sso_config.app_allowAnonymous=app_allowAnonymous;
```
Comment (rushmin) [inline modules/apps/publisher/themes/appm/js/logic/create.asset.js]: Need spaces !

File: modules/apps/publisher/themes/appm/partials/edit-asset.hbs [hunk 9]
```diff
@@ -69,6 +70,18 @@
 								</div>
 							</div>
 
+                         <div class="form-group">
+                             <label class="control-label col-sm-2">Allow Anonymous:  </label>
+                             <div class="col-sm-10">
+								 <label>
```
Comment (rushmin) [inline modules/apps/publisher/themes/appm/partials/edit-asset.hbs]: Formatting issues

File: modules/apps/publisher/themes/appm/partials/edit-asset.hbs [hunk 10]
```diff
@@ -174,6 +189,7 @@
 										<th style="width: 60px;">HTTP Verb</th>
 										<th style="width: 100px;">Throttling Tier</th>
 										<th style="width: 70px;">Skip Throttling</th>
+                                        <th style="width: 70px;">Allow Anonymous Access</th>
```
Comment (rushmin) [inline modules/apps/publisher/themes/appm/partials/edit-asset.hbs]: Formatting issues.

File: modules/apps/store/extensions/assets/webapp/themes/store/js/logic/assets/assets.js [hunk 11]
```diff
@@ -2,19 +2,22 @@ $(function() {
...
-				  var port = $("#assetshttpsPort").val()
-				  location.href = localIP + ":" + port + "/store/login";	
-			  }else{
-				  var assetId = $('#asset').data('id');
-				  $('#modal-login').data('value', assetId);
-				  $("#modal-login").modal('show');
+			  var allowAnonymous = $(this).find("input").val();
+			  if (allowAnonymous.toUpperCase() != "TRUE") {
+				  var ssoEnabled = $('#sso').val();
+				  console.log(ssoEnabled);
```
Comment (rushmin) [inline modules/apps/store/extensions/assets/webapp/themes/store/js/logic/assets/assets.js]: Is this console log needed ?

File: modules/apps/publisher/themes/appm/js/resource-add.js [hunk 12]
```diff
@@ -93,6 +93,22 @@ $( document ).ready(function() {
...
 
 
+    $(".anonymous_checkbox").click(function(){
+        var output = [];
+        $( ".anonymous_checkbox" ).each(function( index ) {
+            if( $(this).is(':checked')){
+                 output.push("TRUE");
+            }
+            else
+            {
```
Comment (dinusha-dilrukshi) [inline modules/apps/publisher/themes/appm/js/resource-add.js]: Use one convention to handle braces.

File: modules/apps/store/extensions/assets/webapp/pages/asset.jag [hunk 13]
```diff
@@ -120,11 +120,26 @@ require('/modules/store.js').exec(function (ctx) {
 
         var defaultApplicationName = "DefaultApplication";
 
-        var enterpriseSubscription = subscriptionService.getSubscription(appInfo, defaultApplicationName, "ENTERPRISE", authenticator.getLoggedInUser().username).subscription; 
+        var enterpriseSubscription;
```
Comment (dinusha-dilrukshi) [inline modules/apps/store/extensions/assets/webapp/pages/asset.jag]: In code line 124-139, we have duplicated same if condition twice here. Also the logic need to re-checked.

File: modules/apps/store/extensions/assets/webapp/pages/asset.jag [hunk 14]
```diff
@@ -202,7 +217,13 @@ require('/modules/store.js').exec(function (ctx) {
     }
 
     function isStoreAdmin(){
-        return hasUserRole("Internal/store-admin");
+        if(authenticator.getLoggedInUser()) {
```
Comment (dinusha-dilrukshi) [inline modules/apps/store/extensions/assets/webapp/pages/asset.jag]: authenticator.getLoggedInUser() has called in few places. Assign it into a variable and re-use.

File: modules/apps/store/extensions/assets/webapp/pages/asset.jag [hunk 15]
```diff
@@ -202,7 +217,13 @@ require('/modules/store.js').exec(function (ctx) {
...
+            return hasUserRole("Internal/store-admin");
```
Comment (dinusha-dilrukshi) [inline modules/apps/store/extensions/assets/webapp/pages/asset.jag]: Make Internal/store-admin role configurable. Move it to store.json.

File: modules/apps/store/extensions/assets/webapp/themes/store/js/logic/asset/overview/overview.js [hunk 16]
```diff
@@ -32,17 +32,19 @@ $(function () {
     initializeUserActivity("page-load", jsonObj);
 
     $("#gatewayURL").on('click', function (e) {
-        var isSubscribed = $('#subscribed').val();
-        if (isSubscribed.toLowerCase() === 'false') {
-            $('#messageModal2').html($('#confirmation-data2').html());
-            $('#messageModal2 h3.modal-title').html(('Resource forbidden'));
-            $('#messageModal2 div.modal-body').html('\n\n' + ('You have not subscribed to this Application.'));
-            $('#messageModal2 a.btn-other').html('OK');
+        if ($('#hdnUsertId').val()!="") {
```
Comment (dinusha-dilrukshi) [inline modules/apps/store/extensions/assets/webapp/themes/store/js/logic/asset/overview/overview.js]: Need spaces.

File: modules/apps/publisher/modules/ext/scripts/webapp.action.save.js (hunk 1 above)
Comment (lahiru823) [inline modules/apps/publisher/modules/ext/scripts/webapp.action.save.js]: Added jira https://wso2.org/jira/browse/APPM-1072
This issue in repeated in many instance and all the scenarios will be fixed in future releases.

File: modules/apps/publisher/themes/appm/js/copy-app.js (hunk 3 above)
Comment (lahiru823) [inline modules/apps/publisher/themes/appm/js/copy-app.js]: Already fixed in the latest code

File: modules/apps/publisher/themes/appm/js/copy-app.js (hunk 4 above)
Comment (lahiru823) [inline modules/apps/publisher/themes/appm/js/copy-app.js]: Already fixed in the latest code

File: modules/apps/publisher/themes/appm/js/edit.asset.js (hunk 5 above)
Comment (lahiru823) [inline modules/apps/publisher/themes/appm/js/edit.asset.js]: Already fixed in the latest code

File: modules/apps/publisher/themes/appm/js/edit.asset.js (hunk 6 above)
Comment (lahiru823) [inline modules/apps/publisher/themes/appm/js/edit.asset.js]: Already fixed in the latest code

File: modules/apps/publisher/themes/appm/partials/edit-asset.hbs (hunk 9 above)
Comment (lahiru823) [inline modules/apps/publisher/themes/appm/partials/edit-asset.hbs]: This code has been removed/changed in the latest code

File: modules/apps/publisher/themes/appm/partials/edit-asset.hbs (hunk 10 above)
Comment (lahiru823) [inline modules/apps/publisher/themes/appm/partials/edit-asset.hbs]: This code has been removed/changed in the latest code

File: modules/apps/store/extensions/assets/webapp/themes/store/js/logic/assets/assets.js (hunk 11 above)
Comment (lahiru823) [inline modules/apps/store/extensions/assets/webapp/themes/store/js/logic/assets/assets.js]: Fixed with commit# https://github.com/wso2/carbon-appmgt/commit/771421431cb5e13218652c43834cc94952ca7e05

File: modules/apps/publisher/themes/appm/js/resource-add.js (hunk 12 above)
Comment (lahiru823) [inline modules/apps/publisher/themes/appm/js/resource-add.js]: This has been fixed with a subsequent commit

File: modules/apps/store/extensions/assets/webapp/pages/asset.jag (hunk 13 above)
Comment (lahiru823) [inline modules/apps/store/extensions/assets/webapp/pages/asset.jag]: This code has been changed in the latest code. (logic/where condition has been removed)

File: modules/apps/store/extensions/assets/webapp/pages/asset.jag (hunk 14 above)
Comment (lahiru823) [inline modules/apps/store/extensions/assets/webapp/pages/asset.jag]: Already changed as;
   var username = null;
    var loggedUser = authenticator.getLoggedInUser();
    if (loggedUser != null) {
        username = loggedUser.username;
    }

File: modules/apps/store/extensions/assets/webapp/themes/store/js/logic/asset/overview/overview.js (hunk 16 above)
Comment (lahiru823) [inline modules/apps/store/extensions/assets/webapp/themes/store/js/logic/asset/overview/overview.js]: Already fixed in the latest code.

File: modules/apps/publisher/modules/ext/scripts/webapp.action.save.js (hunk 2 above)
Comment (lahiru823) [inline modules/apps/publisher/modules/ext/scripts/webapp.action.save.js]: Fixed with commit# https://github.com/wso2/carbon-appmgt/commit/52ff1cb7c51f578dfaa5a10cac841bc6791f13b6

File: modules/apps/publisher/themes/appm/js/logic/create.asset.js (hunk 7 above)
Comment (lahiru823) [inline modules/apps/publisher/themes/appm/js/logic/create.asset.js]: Fixed with commit# https://github.com/wso2/carbon-appmgt/commit/52ff1cb7c51f578dfaa5a10cac841bc6791f13b6

File: modules/apps/publisher/themes/appm/js/logic/create.asset.js (hunk 8 above)
Comment (lahiru823) [inline modules/apps/publisher/themes/appm/js/logic/create.asset.js]: Fixed with commit# https://github.com/wso2/carbon-appmgt/commit/52ff1cb7c51f578dfaa5a10cac841bc6791f13b6

File: modules/apps/store/extensions/assets/webapp/pages/asset.jag (hunk 15 above)
Comment (lahiru823) [inline modules/apps/store/extensions/assets/webapp/pages/asset.jag]: Fixed with commit# https://github.com/wso2/carbon-appmgt/commit/c38592d1d8446a7b5ac8eddd4008fae19bc166f3

**Pull request merged**